File: examples/example-android/polar-sensor-data-collector/app/src/main/java/com/polar/polarsensordatacollector/ui/landing/SensorListAdapter.java
package com.polar.polarsensordatacollector.ui.landing;

import android.content.Context;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.polar.polarsensordatacollector.R;
import com.polar.sdk.api.model.PolarDeviceInfo;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class SensorListAdapter extends RecyclerView.Adapter<SensorListAdapter.ViewHolder> {

    final List<PolarDeviceInfo> dataItems;
    private final ItemSelected itemSelected;

    public SensorListAdapter(List<PolarDeviceInfo> dataItems, ItemSelected selected) {
        this.dataItems = dataItems;
        this.itemSelected = selected;
    }

    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.device_list_item, parent, false);
        return new ViewHolder(v, itemSelected);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.name.setText(dataItems.get(position).getName());
        holder.rssi.setText(dataItems.get(position).getRssi() + "dbm");
        final Context context = holder.itemView.getContext();
        if (dataItems.get(position).isConnectable()) {
            holder.background.setBackgroundColor(context.getColor(R.color.color_connected));
        } else if (!dataItems.get(position).isConnectable()) {
            holder.background.setBackgroundColor(context.getColor(R.color.color_not_connectable));
        } else {
            holder.background.setBackgroundColor(Color.TRANSPARENT);
        }
    }

    @Override
    public int getItemCount() {
        return dataItems.size();
    }

    public interface ItemSelected {
        void itemSelected(@Nullable PolarDeviceInfo info);
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private final TextView name;
        private final TextView rssi;
        private final ItemSelected itemSelected;
        private final LinearLayout background;

        public ViewHolder(View v, ItemSelected itemSelected) {
            super(v);
            v.setOnClickListener(this);
            this.itemSelected = itemSelected;
            this.name = v.findViewById(R.id.device_name);
            this.rssi = v.findViewById(R.id.device_rssi);
            this.background = v.findViewById(R.id.device_item_background);
        }

        @Override
        public void onClick(View view) {
            itemSelected.itemSelected(SensorListAdapter.this.dataItems.get(this.getBindingAdapterPosition()));
        }
    }
}

File: examples/example-android/polar-sensor-data-collector/app/src/main/java/com/polar/polarsensordatacollector/ui/landing/SensorListAdapter.kt
package com.polar.polarsensordatacollector.ui.landing

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.polar.polarsensordatacollector.R
import com.polar.sdk.api.model.PolarDeviceInfo

/**
 * RecyclerView adapter for the sensor selection dialog.
 *
 * Displays two sections:
 *  1. Already-connected devices (sorted alphabetically, shown with a green "CONNECTED" badge)
 *  2. Discovered (not yet connected) devices (sorted by RSSI descending)
 *
 * A separator row with the label "Connected" is injected between the two sections whenever
 * at least one connected device is present in the list.
 */
class SensorListAdapter(
    private val connectedDeviceIds: Set<String>,
    private val onItemSelected: (PolarDeviceInfo?) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    sealed class SensorListItem {
        data class Device(val info: PolarDeviceInfo, val isConnected: Boolean) : SensorListItem()
        data class Separator(val label: String) : SensorListItem()
    }

    private val items: MutableList<SensorListItem> = mutableListOf()

    /** Rebuild the displayed list from a flat list of scanned [PolarDeviceInfo].
     *  Call whenever new scan results arrive. */
    fun updateDevices(raw: List<PolarDeviceInfo>) {
        val connected = raw.filter { it.deviceId in connectedDeviceIds }
            .sortedBy { it.name }
        val available = raw.filter { it.deviceId !in connectedDeviceIds }
            .sortedByDescending { it.rssi }

        items.clear()
        if (connected.isNotEmpty()) {
            items.add(SensorListItem.Separator("Connected"))
            connected.forEach { items.add(SensorListItem.Device(it, isConnected = true)) }
            items.add(SensorListItem.Separator("Available"))
        }
        available.forEach { items.add(SensorListItem.Device(it, isConnected = false)) }
        notifyDataSetChanged()
    }

    // ── ViewHolder types ──────────────────────────────────────────────────────────────────────

    inner class DeviceViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val name: TextView = view.findViewById(R.id.device_name)
        val rssi: TextView = view.findViewById(R.id.device_rssi)
        val badge: TextView = view.findViewById(R.id.device_connected_badge)
        val background: LinearLayout = view.findViewById(R.id.device_item_background)

        init {
            view.setOnClickListener {
                val pos = bindingAdapterPosition
                if (pos != RecyclerView.NO_ID.toInt()) {
                    (items[pos] as? SensorListItem.Device)?.let {
                        onItemSelected(it.info)
                    }
                }
            }
        }
    }

    class SeparatorViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val label: TextView = view.findViewById(R.id.separator_label)
    }

    // ── Adapter overrides ─────────────────────────────────────────────────────────────────────

    companion object {
        private const val VIEW_TYPE_DEVICE = 0
        private const val VIEW_TYPE_SEPARATOR = 1
    }

    override fun getItemViewType(position: Int): Int = when (items[position]) {
        is SensorListItem.Device -> VIEW_TYPE_DEVICE
        is SensorListItem.Separator -> VIEW_TYPE_SEPARATOR
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            VIEW_TYPE_SEPARATOR -> SeparatorViewHolder(
                inflater.inflate(R.layout.device_list_separator, parent, false)
            )
            else -> DeviceViewHolder(
                inflater.inflate(R.layout.device_list_item, parent, false)
            )
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (val item = items[position]) {
            is SensorListItem.Separator -> {
                (holder as SeparatorViewHolder).label.text = item.label
            }
            is SensorListItem.Device -> {
                holder as DeviceViewHolder
                holder.name.text = item.info.name
                holder.rssi.text = "${item.info.rssi} dbm"

                if (item.isConnected) {
                    holder.badge.visibility = VISIBLE
                    holder.background.setBackgroundColor(
                        holder.itemView.context.getColor(R.color.color_already_connected)
                    )
                } else if (item.info.isConnectable) {
                    holder.badge.visibility = GONE
                    holder.background.setBackgroundColor(
                        holder.itemView.context.getColor(R.color.color_connected)
                    )
                } else {
                    holder.badge.visibility = GONE
                    holder.background.setBackgroundColor(Color.TRANSPARENT)
                }
            }
        }
    }

    override fun getItemCount(): Int = items.size

    // Stable IDs: use device ID hash for devices, a fixed sentinel for separators.
    override fun getItemId(position: Int): Long = when (val item = items[position]) {
        is SensorListItem.Device -> item.info.deviceId.hashCode().toLong()
        is SensorListItem.Separator -> item.label.hashCode().toLong() + Long.MIN_VALUE / 2
    }

    // Legacy interface kept for Java interop (DialogUtility is Kotlin so this is for safety).
    fun interface ItemSelected {
        fun itemSelected(info: PolarDeviceInfo?)
    }
}
